app/workers/import/row_worker.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+class Import::RowWorker
+  include Sidekiq::Worker
+
+  sidekiq_options queue: 'pull', retry: 6, dead: false
+
+  sidekiq_retries_exhausted do |msg, _exception|
+    # Increment the total number of processed items, and bump the state of the import if needed
+    bulk_import_id = BulkImportRow.where(id: msg['args'][0]).pick(:id)
+    unless bulk_import_id.nil?
+      # Use `increment_counter` so that the incrementation is done atomically in the database
+      BulkImport.increment_counter(:processed_items, bulk_import_id) # rubocop:disable Rails/SkipsModelValidations
+      bulk_import = BulkImport.find(bulk_import_id)
+      bulk_import.update!(state: :finished, finished_at: Time.now.utc) if bulk_import.processed_items == bulk_import.total_items
+    end
+  end
+
+  def perform(row_id)
+    row = BulkImportRow.eager_load(bulk_import: :account).find_by(id: row_id)
+    return true if row.nil?
+
+    imported = BulkImportRowService.new.call(row)
+
+    mark_as_processed!(row, imported)
+  end
+
+  private
+
+  def mark_as_processed!(row, imported)
+    bulk_import_id = row.bulk_import_id
+    row.destroy! if imported
+
+    # Use `increment_counter` so that the incrementation is done atomically in the database
+    BulkImport.increment_counter(:processed_items, bulk_import_id) # rubocop:disable Rails/SkipsModelValidations
+    BulkImport.increment_counter(:imported_items, bulk_import_id) if imported # rubocop:disable Rails/SkipsModelValidations
+    bulk_import = BulkImport.find(bulk_import_id)
+    bulk_import.update!(state: :finished, finished_at: Time.now.utc) if bulk_import.processed_items == bulk_import.total_items
```

Comment:
So this logic is used in at least two separate places, maybe it should be a method on `BulkImport`, something like:

```ruby
def progress!
  update!(state: :finished, finished_at: Time.now.utc) if processed_items == total_items
end
```

Reply:
Moved it to `BulkImport.progress!`, it's a class method rather than an instance method because it takes care of the atomic increments as well before loading the model.